Add UnbinnedParameterLikelihood

dvandyk opened the pull request:
I have started working on an UnbinnedParameterLikelihood for smelli. Unfortunately, something
is going wrong. I would appreciate some help.

According to the docstring, ```GlobalLikelihood._log_likelihood``` receives an object of type ```wilson.Wilson```. That does not seem to be the case, and some caching methods are missing.

PS: Fixed some typos in a docstring, and added the dependency on wcxf to make the tests pass.

DavidMStraub commented:
https://github.com/smelli/smelli/blob/master/smelli/classes.py#L328
`par_dict` is a dictionary of parameters. This was necessary for the consistent treatment of NP in CKM. Probably the docstring was not adapted. `GlobalLikelihood.par_dict_sm` gives this dictionary in the absence of NP in CKM, `GlobalLikelihoodPoint.par_dict_np` in its presence.

BTW, the name (Unbinned) doesn't seem optimal. Maybe something involving the fact that you provide samples? Precomputed or sampled likelihood? Presampled?

peterstangl commented on smelli/unbinned.py:
```diff
@@ -0,0 +1,78 @@
+        self.scale   = scale
+        self.names   = names
+        self.samples = samples
+        self.weights = weights if weights is None else np.ones(len(samples))
```
Shouldn't this be
```python
self.weights = np.ones(len(samples)) if weights is None else weights
```
?

DavidMStraub replied:
More Pythonic:
```python
self.weights = weights or np.ones(len(samples))
```

dvandyk replied:
Indeed. I had tested only with non-empty weights...

dvandyk replied:
@DavidMStraub your ore pythonic version does not work, since in the deserialization weights is provided as an ```np.array``` which has an implicit conversion to a boolean value (with warning). I will stick to the 'is not None'.

dvandyk commented on smelli/unbinned.py:
```diff
@@ -0,0 +1,78 @@
+    @classmethod
+    def load(cls, f, **kwargs):
+        """Instantiate an object from a YAML string or stream."""
+        d = yaml.load(f, Loader=yaml.Loader)
```
@DavidMStraub @peterstangl I'm overriding the use of the ```SafeLoader``` here since I can dump but not load the ```np.array``` otherwise. Any recommendations?

DavidMStraub replied:
Are you storing an actual numpy array in a YAML file? If it's just an array of numbers, I would suggest using the `tolist()` method and store it as a native YAML (nested) list. This would avoid such problems.

Then again, there is a question of design: is it really a good idea to store all the samples in the YAML? I mean, in practice, this will be a huge file, so wouldn't it make more sense to store it somewhere more convenient (up to the user, `.npy`, HDF5, ...) and use the YAML only for metadata?

dvandyk replied:
@DavidMStraub's advice has been followed. Using .npy files.

dvandyk commented on smelli/unbinned.py:
```diff
@@ -0,0 +1,78 @@
+        schema = self.output_schema()
+        d = schema(d)
+        # remove NoneTypes and empty lists
+        d = {k: v for k, v in d.items() if v is not None}
```
@DavidMStraub @peterstangl I have removed the flavio check whether a list is not the empty list, since in combination with a numpy array this led to removal of n-dimensional arrays.

dvandyk commented:
> https://github.com/smelli/smelli/blob/master/smelli/classes.py#L328
> `par_dict` is a dictionary of parameters. This was necessary for the consistent treatment of NP in CKM. Probably the docstring was not adapted. `GlobalLikelihood.par_dict_sm` gives this dictionary in the absence of NP in CKM, `GlobalLikelihoodPoint.par_dict_np` in its presence.

I will have a look tomorrow. But that does not illuminate my confusion: what type is the second parameter ```w```? It does seem to have the ```wilson.Wilson``` interface, i.e., it has a ```match_and_run``` method. But using it leads to an ```AttributeError```, complaining about the absence of the attribute ```_cache``` in ```w```.

> 
> BTW, the name (Unbinned) doesn't seem optimal. Maybe something involving the fact that you provide samples? Precomputed or sampled likelihood? Presampled?

It is the technical term after all. Instead of using a parametric form of a histogram of the ''data'',
it uses random variates of the data. Where do you see a problem? With the outcome of an
unbinned experimental analysis?

peterstangl commented:
> > https://github.com/smelli/smelli/blob/master/smelli/classes.py#L328
> > `par_dict` is a dictionary of parameters. This was necessary for the consistent treatment of NP in CKM. Probably the docstring was not adapted. `GlobalLikelihood.par_dict_sm` gives this dictionary in the absence of NP in CKM, `GlobalLikelihoodPoint.par_dict_np` in its presence.
> 
> I will have a look tomorrow. But that does not illuminate my confusion: what type is the second parameter `w`? It does seem to have the `wilson.Wilson` interface, i.e., it has a `match_and_run` method. But using it leads to an `AttributeError`, complaining about the absence of the attribute `_cache` in `w`.

I do not really understand how this can happen since a `wilson.Wilson` instance should actually have an attribute `_cache` (cf. https://github.com/wilson-eft/wilson/blob/master/wilson/classes.py#L125). So do I understand you correctly that the `wilson.Wilson` instance you are passing to `GlobalLikelihood._log_likelihood` does not have such an attribute?

DavidMStraub commented:
@dvandyk, can you provide a minimal example of what actually fails? I guess we don't understand the problem.

> It is the technical term after all. Instead of using a parametric form of a histogram of the ''data'',
it uses random variates of the data. Where do you see a problem? With the outcome of an
unbinned experimental analysis?

I would suggest some name that allows to infer how this differs from all the rest that is currently implemented in smelli, namely likelihoods that are constructed by plugging WCs into observables and plugging the observables into experimental likelihoods.

In some sense, what you are doing is precomputing a sublikelihood in some way and providing it in the form of random samples (i.e. their density trace the likelihood).  This *could* come from an experimental unbinned analysis, but it could also be used as an intermediate step with any kind of underlying data (binned or not). No?

dvandyk commented:
> @dvandyk, can you provide a minimal example of what actually fails? I guess we don't understand the problem.

Will do, once the version with external files works.

> 
> > It is the technical term after all. Instead of using a parametric form of a histogram of the ''data'',
> > it uses random variates of the data. Where do you see a problem? With the outcome of an
> > unbinned experimental analysis?
> 
> I would suggest some name that allows to infer how this differs from all the rest that is currently implemented in smelli, namely likelihoods that are constructed by plugging WCs into observables and plugging the observables into experimental likelihoods.

Well, in this case the pseudo-observable is a WET Wilson coefficient. So in this respect the new likelihood is not different. What is different is that this has no expectation of being a Gaussian likelihood. Instead, it returns the sum of absolute distances squared between the predicted value
of the vector of WET Wilson coefficients and the each sample.

> 
> In some sense, what you are doing is precomputing a sublikelihood in some way and providing it in the form of random samples (i.e. their density trace the likelihood). This _could_ come from an experimental unbinned analysis, but it could also be used as an intermediate step with any kind of underlying data (binned or not). No?

Correct. But the technical term is still an unbinned likelihood, as in: one has to perform an unbinned likelihood fit to obtain the constraints in the space of SMEFT WCs. Also, there is no usable test statistic for this likelihood, i.e., not pull and no chi^2.

DavidMStraub commented:
Nitpicking: at present the likelihood is not necessarily Gaussian either, only for the `FastLikelihood` instances.

I guess my problem is that "unbinned" is a technical term mostly in the realm of experimental HEP to distinguish it from the usual binned analyses.

Anyway, let's make it work before deciding about names.

dvandyk commented:
Updated the branch.

@peterstangl the reason for my trouble with ```match_run``` was that ```w``` was an instance of ```flavio.WilsonCoefficients``` within the context of ```log_likelihood_sm```, but an instance of ```wilson.Wilson``` otherwise.

@DavidMStraub added a small cli to list and download data sets.

The following example now works:
```
import smelli
gl = smelli.GlobalLikelihood(include_likelihoods=('unbinned_flavio_test',))
display(gl.unbinned_likelihoods)
glp = gl.parameter_point({'lq1_2223': 1e-8}, scale=1000)
display(glp.log_likelihood_dict())
```

dvandyk commented:
This looks good, I still need to remove the test case with the old datafile, and make sure that the new test cases can run by downloading the dataset when running the test case.

DavidMStraub commented:
NB, we first have to release smelli v2.0 in the next few days before merging this.

dvandyk commented:
@DavidMStraub Sure! I don't think this is ready yet for merging, anyway. After discussing with Dan and Marcin yesterday, I think I should use a proper kernel density estimator. Looking for a fast python-based one now. I think Fred had a good candidate...

dvandyk commented:
@DavidMStraub The one that Fred and I had originally used with EOS cannot be automatically installed with Python. (It's called 'figtree', see [Fred's wrapper](https://github.com/fredRos/pyfigtree)).

Following the discussion [here](https://jakevdp.github.io/blog/2013/12/01/kernel-density-estimation/) I would suggest to use [scikit-learn](https://scikit-learn.org/stable/modules/density.html).

dvandyk commented:
sklearn's ```KernelDensity``` does not support complex numbers. I have added a workaround that
converts a numpy array of complex numbers into a numpy array of real numbers, which queries ```wcxf.Bases```. Would appreciate if somebody had a critical look at that.

I also added a test of the class by adapting to the first half of the samples, and computing the perplexity of the second half of samples.

peterstangl commented on smelli/unbinned.py:
```diff
@@ -0,0 +1,111 @@
+                self.accessors.append(lambda wcs: wcs[i])
+            else:
+                self.accessors.append(lambda wcs: np.real(wcs[i]))
+                self.accessors.append(lambda wcs: np.imag(wcs[i]))
```
@dvandyk I think this doesn't work. `i` should be overwritten in each iteration such that in the end, `i` will be just `len(names)-1`. So I think all functions in `self.accessors` would use this final `i` when they are evaluated in the following and thus all values in each row will be equal to the value of the last Wilson coefficient of that row.

peterstangl replied:
Another issue might be that `if 'real' in basis_wcs[n]`, `wcs[i]` is just copied without turning it to a real number. So if it stems from a numpy array of complex type, it's imaginary part might just be `1e-16` but it would still be of complex type. This then will make any numpy array constructed with this number as an entry complex again. I don't know if this could be a problem for the kde. So I would also use `np.real()` in this case just to be sure that no complex numbers remain.

dvandyk replied:
> Another issue might be that `if 'real' in basis_wcs[n]`, `wcs[i]` is just copied without turning it to a real number. ...

Acknowledged and changed locally. Thanks Peter!

dvandyk replied:
> @dvandyk I think this doesn't work. `i` should be overwritten in each iteration such that in the end, `i` will be just `len(names)-1`. So I think all functions in `self.accessors` would use this final `i` when they are evaluated in the following and thus all values in each row will be equal to the value of the last Wilson coefficient of that row.

Peter, thank you very much! I tested and confirmed it at the hand of this snippet:
```
sin = ['m', 'l', 'i', 'l', 'e', 's']
idx = [5, 0, 4, 1, 3, 2]
acc = []
for i in idx:
    acc.append(lambda mys: mys[i])

sout = [a(sin) for a in acc]
sout
```
The output is indeed ```['i', 'i', 'i', 'i', 'i', 'i']```. This is completely unexpected for me,
and does not follow the idea of a lambda functional at all, but that's a Python issue.

It works if I return the lambda from a different function, though. Using that for the time being.

peterstangl commented on smelli/unbinned.py:
```diff
@@ -0,0 +1,111 @@
+
+        self.samples = np.array([
+            [a(row) for a in self.accessors] for row in self.samples
+        ])
```
Instead of using the accessors (which might not work, see my comment above), one could do something like this:

Outside of class, import chain function
```python
from itertools import chain
```

Inside `__init__` method of `UnbinnedParameterLikelihood` class, make `basis_wcs` a private attribute
```python
self._basis_wcs = basis_wcs
```

Inside `UnbinnedParameterLikelihood` class, define utility method for converting list of n complex values and m real values to list of 2n+m real values given that `self.names` contains n+m names that can be found in `self._basis_wcs`
```python
def _complex_to_real(self, values):
    return list(chain.from_iterable(
        [np.real(values[i])] if 'real' in self._basis_wcs[n]
        else
        [np.real(values[i]), np.imag(values[i])]
        for i, n in enumerate(self.names)
    ))
```
convert the samples using
```python
self.samples = np.array([
    self._complex_to_real(row) for row in self.samples
])
```
Use the `_complex_to_real` method also later inside the ` log_likelihood` method on `values`:
```python
values = self._complex_to_real([wet_wc[name] for name in self.names])
```

dvandyk replied:
Sorry Peter, I did not see this comment before just now, after fixing my approach. Reading up on ```chain``` right now.

peterstangl replied:
`chain` can be used to flatten a list:
```python
list(chain.from_iterable([[1], [2,3], [4,5], [6], [7]]))
```
returns
```
[1, 2, 3, 4, 5, 6, 7]
```
So one can put real values in single-element lists and complex values in two-element lists and then flatten the list using `chain`.

dvandyk commented:
Ok, I fixed the lambda problem following Peter's suggestion. Thanks a lot @peterstangl !

peterstangl commented on smelli/unbinned.py:
```diff
@@ -0,0 +1,167 @@
+        else:
+            smeft_wc = Wilson.from_wc(w.get_wcxf(sector='all', scale=self.scale, par=par_dict, eft='SMEFT', basis='Warsaw'))
+        wet_wc = smeft_wc.match_run(self.scale, self.eft, self.basis)
+        values = _complex_to_real([wet_wc[name] for name in self.names])
```
@dvandyk don't you need a `self.` here in front of `_complex_to_real`? The list comprehension you are using in this line could actually be avoided by modifying the `_complex_to_real` method a bit to allow for direct input of a dictionary. This could e.g. be done like

in `__init__`, after the definition of `self._basis_wcs`, define
```python
self._wc_is_real = {n:'real' in self._basis_wcs[n] for n in self.names}
```
use an updated `_complex_to_real` method that can take a dictionary as argument:
```python
def _complex_to_real(self, values):
    if isinstance(values, dict):
        iterator = self._wc_is_real.items()
    else:
        iterator = enumerate(self._wc_is_real.values())
    return list(chain.from_iterable(
        [np.real(values[i])] if real
        else
        [np.real(values[i]), np.imag(values[i])]
        for i, real in iterator
    ))
```

And then the above line could just look like
```python
values = self._complex_to_real(wet_wc)
```

peterstangl replied:
@dvandyk, sorry, of course the dictionary `self._wc_is_real` has to be ordered. So one needs to use an `OrderedDict`:
```python
from collections import OrderedDict
```
and then
```python
self._wc_is_real = OrderedDict((n, 'real' in self._basis_wcs[n]) for n in self.names)
```
